src/components/ui/OakTextInput/OakTextInput.tsx

```diff
@@ -0,0 +1,196 @@
+import React, { ChangeEventHandler } from "react";
+import styled from "styled-components";
+
+import { parseColor } from "@/styles/helpers/parseColor";
+import {
+  InternalTextInput,
+  InternalTextInputProps,
+} from "@/components/base/InternalTextInput";
+import { OakFlex, OakIcon, OakIconName } from "@/components/base";
+import { OakCombinedColorToken, OakDropShadowToken } from "@/styles";
+import { parseDropShadow } from "@/styles/helpers/parseDropShadow";
+
+type StyledTextInputWrapperProps = {
+  $color: OakCombinedColorToken;
+  $hoverBackground: OakCombinedColorToken;
+  $background: OakCombinedColorToken;
+  $borderColor: OakCombinedColorToken;
+  $focusRingDropShadows: OakDropShadowToken[];
+  $disabledBackgroundColor: OakCombinedColorToken;
+  $readOnlyBorderColor: OakCombinedColorToken;
+  $disabledColor: OakCombinedColorToken;
+  $readOnlyColor: OakCombinedColorToken;
+};
+
+export type OakTextInputProps = {
+  id?: string;
+  type?: "text" | "password" | "number" | "email" | "tel";
+  /**
+   * Disables user input and updates the appearance accordingly.
+   */
+  disabled?: boolean;
+  /**
+   * Makes the input read-only. Preventing the user from changing the value.
+   */
+  readOnly?: boolean;
+  /**
+   * Sets the value. Use this in controlled components;
+   */
+  value?: string;
+  /**
+   * Sets the initial value. Use this for an uncontrolled component;
+   */
+  defaultValue?: string;
+  /**
+   * Used to target the input element in tests.
+   */
+  "data-testid"?: string;
+  placeholder?: string;
+  onChange?: ChangeEventHandler<HTMLInputElement>;
+  /**
+   * Alters the appearance of the input field to indicate whether the input is valid or invalid.
+   */
+  validity?: "valid" | "invalid";
+  /**
+   * Displays an icon at the start of the input.
+   */
+  startEnhancerIconName?: OakIconName;
+  /**
+   * Displays an icon at the end of the input.
+   */
+  endEnhancerIconName?: OakIconName;
+  $iconColor?: OakCombinedColorToken;
+  $validBorderColor?: OakCombinedColorToken;
+  $invalidBorderColor?: OakCombinedColorToken;
+  $validIconColor?: OakCombinedColorToken;
+  $invalidIconColor?: OakCombinedColorToken;
+} & Partial<StyledTextInputWrapperProps> &
+  Pick<InternalTextInputProps, "$width" | "$maxWidth" | "onInitialFocus">;
+
+const StyledTextInputWrapper = styled(OakFlex)<StyledTextInputWrapperProps>`
+  &:hover {
+    cursor: text;
+  }
+
+  &:focus-within:not(:has(input:disabled, input:read-only)) {
+    box-shadow: ${(props) =>
+      props.$focusRingDropShadows
+        .map((dropShadow) => parseDropShadow(dropShadow))
+        .join(",")};
+  }
+
+  background: ${(props) => parseColor(props.$background)};
+
+  @media (hover: hover) {
+    &:hover:not(:focus-within, :has(input:read-only)) {
+      background: ${(props) => parseColor(props.$hoverBackground)};
+    }
+  }
+
+  &:has(input:read-only) {
```

Comment:
I've seen use of `:has` elsewhere, are we okay targeting only evergreen browsers?

Reply:
Looking at https://caniuse.com/css-has versus https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices#browsers-to-test-in-from-june-2022, I think it's worth exploring potential polyfills* or alternative solutions here.

*we'd need to targeting Firefox `> v103` (desktop) and Safari `> 15.4` (desktop and mobile)

Reply:
... but it depends if the native fallback is functional and accessible (or not)

Reply:
given the styling of requirements of the quiz checkbox an alternative solution to :has would be a pretty heavy rewrite and we're already behind. Polyfill sounds like a a good idea to me. Looking at your above comment James it seems like the main issue would be safari iOS. I'd still put this on the backlog.

Reply:
Okay, I think is now moot for `OakInput` at least. The only a11y issue I think it introduced was the lack of a focus ring in unsupported browsers. However I was only using `:has` to disable the focus styles when the input was `readOnly` — in retrospect this was incorrect. As per https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly#attribute_interactions a `readOnly` input should still receive focus.